File: tnzhou/56.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    static bool cmp(Interval a, Interval b) {
        return a.start < b.start;
    }
    
    vector<Interval> merge(vector<Interval>& intervals) {
        sort(intervals.begin(), intervals.end(), cmp);
        int current_left = -1;
        int current_right = -1;
        vector<Interval> result;
        if (intervals.size() == 0) {
            return result;
        }
        current_left = intervals[0].start;
        current_right = intervals[0].end;
        for (int i = 1; i < intervals.size(); i ++) {
            Interval itv = intervals[i];
            if (itv.start <= current_right && itv.end >= current_left) {
                current_left = min(itv.start, current_left);
                current_right = max(itv.end, current_right);
            } else {
                result.push_back(Interval(current_left, current_right));
                current_left = itv.start;
                current_right = itv.end;
            }
        }
        result.push_back(Interval(current_left, current_right));
        return result;
    }
};

File: chunlin/56.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        result = list()
        for interval in sorted(intervals, key=lambda i: i.start):
            if result and result[-1].end >= interval.start:
                result[-1].end = max(result[-1].end, interval.end)  # result[-1].end = interval.end
            else:
                result.append(interval)
        return result

    
    # WA: [[1,4],[2,3]] -> Output:[[1,3]] Expected:[[1,4]]
